### PR TITLE
Ensure that known data chunk deletions do not initiate background deletes

### PR DESCRIPTION
**Pull request by pnarayanan**
Ensure that chunk deletions initiated by operations such as composite
blob deletions do not attempt any more background chunk deletions. Given
that they are known to be data chunks, the Get requests are unnecessary.

Also background chunk deletion operations are expected to fail from time to
time in the presence of duplicate deletes and such. Therefore, use trace logging
to log these errors and not error logging.

### CODE REVIEW

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 1]
```diff
@@ -281,24 +278,25 @@ public void onCompletion(Void result, Exception exception) {
...
-            deleteRequests.add(new BackgroundDeleteRequest(storeKey, serviceId));
-          }
-          initiateBackgroundDeletes(deleteRequests);
+    Callback<GetBlobResultInternal> callback = (GetBlobResultInternal result, Exception exception) -> {
+      if (exception != null) {
+        // It is expected that these requests will not always succeed. For example, this may have been triggered by a
+        // duplicate delete and the blob could have already been hard deleted, so the deserialization can fail, or the
+        // blob could have been garbage collected and not found at all and so on.
+        logger.trace(
+            "Encountered exception when attempting to get chunks of a possibly composite deleted blob " + blobId,
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: use log formatting

File: ambry-router/src/test/java/com.github.ambry.router/MockServer.java [hunk 2]
```diff
@@ -63,6 +64,7 @@
   private boolean getErrorOnDataBlobOnly = false;
   private final ClusterMap clusterMap;
   private final String dataCenter;
+  private final HashMap<RequestOrResponseType, Integer> requestCounts = new HashMap<>();
```
Comment (cgtz) [inline ambry-router/src/test/java/com.github.ambry.router/MockServer.java]: should this be threadsafe (i.e. ConcurrentMap and AtomicInteger)? I guess the test where it is used does not include any multi-threading

Comment (pnarayanan) [inline ambry-router/src/test/java/com.github.ambry.router/MockServer.java]: Yeah, I didn't think this needs to be concurrent - but just made it one to be safe.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 3]
```diff
@@ -281,24 +278,25 @@ public void onCompletion(Void result, Exception exception) {
...
+            exception);
+      } else if (result.getBlobResult != null) {
+        logger.error("Unexpected result returned by background get operation to fetch chunk ids.");
+      } else if (result.storeKeys != null) {
+        List<BackgroundDeleteRequest> deleteRequests = new ArrayList<>(result.storeKeys.size());
+        for (StoreKey storeKey : result.storeKeys) {
+          logger.trace("Initiating delete of chunk blob: " + storeKey);
```
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: fixed.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 1 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: fixed.

**Comment by pnarayanan** [comment]
@vgkholla @cgtz updated.

File: ambry-router/src/test/java/com.github.ambry.router/MockServer.java [hunk 4]
```diff
@@ -64,7 +63,7 @@
   private boolean getErrorOnDataBlobOnly = false;
   private final ClusterMap clusterMap;
   private final String dataCenter;
-  private final HashMap<RequestOrResponseType, Integer> requestCounts = new HashMap<>();
+  private final ConcurrentHashMap<RequestOrResponseType, Integer> requestCounts = new ConcurrentHashMap<>();
```
Comment (cgtz) [inline ambry-router/src/test/java/com.github.ambry.router/MockServer.java]: Is AtomicInteger required here?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 5]
```diff
@@ -281,24 +278,24 @@ public void onCompletion(Void result, Exception exception) {
...
+        // duplicate delete and the blob could have already been hard deleted, so the deserialization can fail, or the
+        // blob could have been garbage collected and not found at all and so on.
+        logger.trace("Encountered exception when attempting to get chunks of a possibly composite deleted blob {} ",
+            blobId, exception);
+      } else if (result.getBlobResult != null) {
+        logger.error("Unexpected result returned by background get operation to fetch chunk ids.");
+      } else if (result.storeKeys != null) {
+        List<BackgroundDeleteRequest> deleteRequests = new ArrayList<>(result.storeKeys.size());
+        for (StoreKey storeKey : result.storeKeys) {
+          logger.trace("Initiating delete of chunk blob {}: ", storeKey);
```
Comment (cgtz) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: switch the `{}` and `:`

Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: fixed.

File: ambry-router/src/test/java/com.github.ambry.router/MockServer.java (hunk 4 above)
Comment (pnarayanan) [inline ambry-router/src/test/java/com.github.ambry.router/MockServer.java]: I see what you mean. Fixed this using `LongAdder` (supposedly better than `AtomicLong` for this purpose - not that it matters much here).

**Comment by pnarayanan** [comment]
@cgtz updated.

**Comment by cgtz** [comment]
LGTM, @vgkholla can I go ahead and merge, or do you want to take another look?

**Pull request merged**